enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/net/ServerStateTest.java

```diff
@@ -0,0 +1,127 @@
+import org.neo4j.causalclustering.helper.EnableableLifeCylcleEnableableStateChangeTest;
+import org.neo4j.helpers.ListenSocketAddress;
+import org.neo4j.logging.FormattedLogProvider;
+import org.neo4j.logging.Level;
+import org.neo4j.ports.allocation.PortAuthority;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * More generalized state tests of EnableableLifeCycle can be found {@link EnableableLifeCycleStateChangeTest} and
+ * {@link EnableableLifeCylcleEnableableStateChangeTest}
+ */
+public class ServerStateTest
+{
+    private Server server;
+    private final EventLoopGroup clientGroup = new NioEventLoopGroup();
+
+    @Before
+    public void setUp() throws Throwable
+    {
+        server = createServer();
+        server.init();
+        assertFalse( canConnect( server.address(), clientGroup ) );
+    }
+
+    @After
+    public void tearDown() throws Throwable
+    {
+        server.stop();
+        server.shutdown();
+    }
+
+    @Test
+    public void shouldStartServerNormally() throws Throwable
+    {
+        server.start();
+        assertTrue( canConnect( server.address(), clientGroup ) );
+    }
+
+    @Test
+    public void canDisableAndEnableServer() throws Throwable
+    {
+        server.start();
+        assertTrue( canConnect( server.address(), clientGroup ) );
+
+        server.disable();
+        assertFalse( canConnect( server.address(), clientGroup ) );
+
+        server.enable();
+        assertTrue( canConnect( server.address(), clientGroup ) );
+    }
+
+    @Test
+    public void serverCannotBeEnabledIfLifeCycleHasNotStarted() throws Throwable
+    {
+        server.enable();
+        assertFalse( canConnect( server.address(), clientGroup ) );
+
+        server.start();
+        assertTrue( canConnect( server.address(), clientGroup ) );
+    }
+
+    @Test
+    public void serverCannotStartIfDisabled() throws Throwable
+    {
+        server.disable();
+
+        server.start();
+        assertFalse( canConnect( server.address(), clientGroup ) );
+
+        server.enable();
+        assertTrue( canConnect( server.address(), clientGroup ) );
+    }
+
+    private static Server createServer()
+    {
+        return new Server( channel ->
+                           {
+                           }, FormattedLogProvider.withDefaultLogLevel( Level.DEBUG ).toOutputStream( System.out ),
+                           FormattedLogProvider.withDefaultLogLevel( Level.DEBUG ).toOutputStream( System.out ),
+                           new ListenSocketAddress( "localhost", PortAuthority.allocatePort() ), "serverName" );
+    }
+
+    private static boolean canConnect( ListenSocketAddress socketAddress, EventLoopGroup eventExecutors ) throws InterruptedException
+    {
+        return new Bootstrap().group( eventExecutors ).channel( NioSocketChannel.class ).handler( new ChannelInitializer<NioSocketChannel>()
```

Comment:
Could the same `Bootstrap` be used across tests, rather than initialising each time?